lib/pbench/agent/tool_meister_client.py

```diff
@@ -87,149 +79,120 @@ def __init__(
             )
+        sig_pub.initialize_and_wait(
+            1,
+            list(tm_allowed_actions),
+            tag="from_pbench_client",
+        )
+        self.sig_pub = sig_pub
+        self.logger.debug("constructed SignalExporter() object")
         return self
 
     def __exit__(self, *args):
-        """On context exit, just close down the to client channel object."""
-        if self.to_client_chan is not None:
-            self.to_client_chan.close()
+        """On context exit, just close down the to SignalExporter object."""
+        self.sig_pub.shutdown()
 
-    def publish(self, group, directory, action, args=None):
-        """publish - marshal a JSON document formed from the group, directory,
-        action, and args arguments.
+    def publish(self, group: str, directory: str, action: str, args=None) -> int:
+        """Publish a state signal formed from the group, directory, action,
+        and args arguments, using the SignalExporter instance.  It waits for
+        responses from subscribers (TDS, and the TMs indirectly) before
+        continuing.
 
         Returns 0 on success, 1 on failure; logs are also written for any
         errors encountered.
+
         """
-        if action not in api_tm_allowed_actions:
+        if action not in tm_allowed_actions:
+            self.logger.warning(f"Attempted to publish illegal action '{action}'")
             return 1
+
         # The published message contains four pieces of information:
         #   {
-        #     "action": "< 'start' | 'stop' | 'send' | 'kill' >",
+        #     "action": "< 'init' | 'start' | 'stop' | 'send' | 'end' | 'terminate' >",
         #     "group": "< the tool group name for the tools to operate on >",
         #     "directory": "< the local directory path to store collected data >"
         #     "args": "< arbitrary argument payload for a particular action >"
         #   }
         # The caller of tool-meister-client must be sure the directory argument
         # is accessible by the Tool Data Sink instance.
-        self.logger.debug("publish %s on chan %s", action, self.from_client_channel)
-        msg = dict(action=action, group=group, directory=str(directory), args=args)
+
+        metadata = dict(
+            group=group,
+            # This check/conversion is necessary, as the pbench scripts
+            # sometimes pass in the directory as a str, and sometimes as
+            # pathlib.PosixPath
+            directory=None if directory is None else str(directory),
+            args=args,
+        )
+        self.logger.debug(
+            "publish state signal for state %s with metadata: %s", action, metadata
+        )
         try:
-            num_present = self.redis_server.publish(
-                self.from_client_channel, json.dumps(msg)
+            resp, msgs = self.sig_pub.publish_signal(
+                event=action, tag="from_pbench_client", metadata=metadata, timeout=100
             )
```

Comment:
That `100` looks like a "magic number".  It would be good to have it rendered as a symbolic constant, for all the usual reasons.